### PR TITLE
Hide download and email if the certificate has been signed by scep request

### PR DESCRIPTION
**Pull request by fdurand**
# Description
Hide download and email if the certificate has been signed by scep request

# Impacts
PKI

# Delete branch after merge
YES

# Checklist
- [ ] Document the feature
- [ ] Add unit tests
- [ ] Add acceptance tests (TestLink)

# NEWS file entries

## Enhancements
* Hide download and email for a certificate that has been signed by a scep request

### CODE REVIEW

File: go/caddy/pfpki/models/models.go [hunk 1]
```diff
@@ -517,8 +518,8 @@ func (c CA) Put(cn string, crt *x509.Certificate, options ...string) error {
 	if CaDB := c.DB.First(&ca, profiledb[0].CaID).Find(&ca); CaDB.Error != nil {
 		c.DB.First(&ca)
 	}
-
-	if err := c.DB.Create(&Cert{Cn: cn, Ca: ca, CaName: ca.Cn, ProfileName: profiledb[0].Name, SerialNumber: crt.SerialNumber.String(), Mail: attributeMap["emailAddress"], StreetAddress: attributeMap["streetAddress"], Organisation: attributeMap["O"], OrganisationalUnit: attributeMap["OU"], Country: attributeMap["C"], State: attributeMap["ST"], Locality: attributeMap["L"], PostalCode: attributeMap["emailAddress"], Profile: profiledb[0], Key: "", Cert: publicKey.String(), ValidUntil: crt.NotAfter}).Error; err != nil {
+	vrai := true
+	if err := c.DB.Create(&Cert{Cn: cn, Ca: ca, CaName: ca.Cn, ProfileName: profiledb[0].Name, SerialNumber: crt.SerialNumber.String(), Mail: attributeMap["emailAddress"], StreetAddress: attributeMap["streetAddress"], Organisation: attributeMap["O"], OrganisationalUnit: attributeMap["OU"], Country: attributeMap["C"], State: attributeMap["ST"], Locality: attributeMap["L"], PostalCode: attributeMap["emailAddress"], Profile: profiledb[0], Key: "", Cert: publicKey.String(), ValidUntil: crt.NotAfter, Scep: &vrai}).Error; err != nil {
```
Comment (nqb) [inline go/caddy/pfpki/models/models.go]: Can we find another name for this variable :-) ?

**Review by nqb — CHANGES_REQUESTED** [review]

**Review by nqb — CHANGES_REQUESTED** [review]
1. Do we really want to prevent download of SCEP certificates from web admin ?

IMHO, we should allow admin users to download SCEP certificates from web admin because web admin doesn't display all informations of certificates (OCSP URL, key usage, etc.) and if you need these informations, you will have to look into DB.

Certificates generated through SCEP are stored in DB without private keys so it's not a problem. Perhaps we can have a "Copy Certificate" button like we have for CA certificates.

2. Download button is still available when you click on a SCEP certificate

**Comment by fdurand** [comment]
Just added a way to download the public key.
Also you need to recompile the pki in order to have the Download button removed from the admin gui.

**Review by nqb — CHANGES_REQUESTED** [review]
- Existing SCEP certs will have a wrong value of `scep` field in DB. We need to update existing SCEP certs if we are able to distinguish them but I'm not sure it's possible.

**Comment by nqb** [comment]
> * Existing SCEP certs will have a wrong value of `scep` field in DB. We need to update existing SCEP certs if we are able to distinguish them but I'm not sure it's possible.

I don't see an automated way to identify SCEP certificate in DB before.

EDIT: it means that "Download" button will be available for old SCEP certificates which can lead to confusion.

EDIT: fixed by 7153f0551f7d4fd0c25cfaf9269b94ad351b3d2a

**Comment by nqb** [comment]
Just did new tests:
1. Download button is still available for SCEP certificates when you click on a SCEP certificate (not on the listing, only "Copy certificate")
1. SCEP filter doesn't work:
image

=> I expect to be able to select SCEP

@satkunas, could you take a look ?

**Comment by satkunas** [comment]
> Just did new tests:
> 
> 1. Download button is still available for SCEP certificates when you click on a SCEP certificate (not on the listing, only "Copy certificate")
> 2. SCEP filter doesn't work:
>    image
> 
> => I expect to be able to select SCEP
> 
> @satkunas, could you take a look ?

@nqb the `scep` column has been pinned to ensure that the scep column is returned with each item in order to show/hide the download/email buttons. If this column is not showing up you'll need to clear your `pkiCerts::defaultSearch` preference.

File: db/upgrade-X.X-X.Y.sql [hunk 2]
```diff
@@ -47,6 +47,13 @@ DELIMITER ;
 call ValidateVersion;
 DROP PROCEDURE IF EXISTS ValidateVersion;
 
+\! echo "altering pki_certs"
+ALTER TABLE pki_certs
+    ADD COLUMN IF NOT EXISTS `scep` BOOLEAN DEFAULT FALSE AFTER ip_addresses;
+
+\! echo "set pki_certs.scep to true if private key is empty"
+UPDATE TABLE pki_certs
```
Comment (nqb) [inline db/upgrade-X.X-X.Y.sql]:
```suggestion
UPDATE pki_certs
```

**Review by nqb — CHANGES_REQUESTED** [review]

**Comment by nqb** [comment]
Work as expected after I remove my preferences and clean my web browser cache.

**Pull request merged**